### PR TITLE
Hyspec spin incoherent polarization

### LINKED ISSUE #21020: Polarization correction for HYSPEC
In the polarization data analysis, the spin incoherent term can be expressed in terms of non spin-flip intensities measured with polarization 90 degrees apart. However, if the angle `alpha` between momentum transfer and polarization is not 0, one needs to divide the data by `cos(2*aplha)`. See for example formula 10 in [Schweika](http://iopscience.iop.org/article/10.1088/1742-6596/211/1/012026/pdf)

**Note:** the formula is derived for scattering only in the horizontal plane. HYSPEC is close to this case.

### PR DESCRIPTION
**Pull request by AndreiSavici**
New algorithm as part of the workflow for polarized analysis on HYSPEC

**To test:**

Code review. See that the formula implemented is the same as in the documentation

Fixes #21020. 

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [x] Are the unit tests small and test the class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by AndreiSavici** [comment]
Still needs some release notes

**Review by soininen — CHANGES_REQUESTED** [review]
In addition to the few style issues, I think the algorithm could be made agnostic of the instrument's reference frame. Also, the common calculations for both histogram and event data could be refactored to separate function(s) to avoid code duplication.

File: Framework/Algorithms/src/HyspecScharpfCorrection.cpp [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+  inputWS = this->getProperty("InputWorkspace");
+  outputWS = this->getProperty("OutputWorkspace");
+  angle = getProperty("PolarizationAngle");
+  angle *= M_PI / 180.;
+  precision = getProperty("Precision");
+  if (inputWS->run().hasProperty("Ei")) {
+    Ei = boost::lexical_cast<double>(inputWS->run().getProperty("Ei")->value());
+  } else {
+    throw std::invalid_argument(
+        "No Ei value has been set or stored within the run information.");
```
[the remark this answers is not in the document]
Comment (AndreiSavici) [reply Framework/Algorithms/src/HyspecScharpfCorrection.cpp]: It is already at the beginning of exec() function

File: Framework/Algorithms/src/HyspecScharpfCorrection.cpp [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    // Copy the energy transfer axis
+    outputWS->setSharedX(i, inputWS->sharedX(i));
+
+    prog.report();
+    // continue if no detectors, if monitor, or is masked
+    if ((!spectrumInfo.hasDetectors(i)) || spectrumInfo.isMonitor(i) ||
+        spectrumInfo.isMasked(i)) {
+      continue;
+    }
+    // get detector info and calculate the in plane angle
```
[the remark this answers is not in the document]
Comment (AndreiSavici) [reply Framework/Algorithms/src/HyspecScharpfCorrection.cpp]: what typo? I am missing that

File: Framework/Algorithms/src/HyspecScharpfCorrection.cpp [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    Mantid::Kernel::V3D detPos = spectrumInfo.position(i);
+    double thetaPlane =
+        std::atan2((detPos - samplePos).X(), (detPos - samplePos).Z());
```
[the remark this answers is not in the document]
Comment (AndreiSavici) [reply Framework/Algorithms/src/HyspecScharpfCorrection.cpp]: I like this :)

File: Framework/Algorithms/src/HyspecScharpfCorrection.cpp [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+template <class T>
+void HyspecScharpfCorrection::ScharpfEventHelper(std::vector<T> &wevector,
+                                                 double thPlane) {
+  typename std::vector<T>::iterator it;
+  for (it = wevector.begin(); it < wevector.end();) {
+    double Ef;
+    Ef = Ei - it->tof();
+    if (Ef <= 0) {
+      it = wevector.erase(it);
```
[the remark this answers is not in the document]
Comment (AndreiSavici) [reply Framework/Algorithms/src/HyspecScharpfCorrection.cpp]: The answer is that I have no idea. The event list might or might not be sorted

File: docs/source/algorithms/HyspecScharpfCorrection-v1.rst [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+This algorithm calculates the Scharpf angle for every event or energy transfer bin,
+then divides the intensity by :math:`F=\cos^2\alpha-\sin^2\alpha=\cos(2\alpha)`. 
+In places where *F* is less than the `Precision`, the intensity of the output workspace is set to 0.
+
+For a detector at angle :math:`\theta` in the horizontal plane, the angle
+between :math:`Q` and :math:`k_i` is given by
+
+.. math::
+    
+    \gamma=\arctan2\left(-\frac{k_f}{k_i}\sin\theta, 1-\frac{k_f}{k_i}\cos\theta\right)
```
[the remark this answers is not in the document]
Comment (AndreiSavici) [reply docs/source/algorithms/HyspecScharpfCorrection-v1.rst]: It's important that it's arctan2. The `arctan` function is only between plus and minus Pi/2

File: Framework/Algorithms/src/HyspecScharpfCorrection.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (soininen) [reply Framework/Algorithms/src/HyspecScharpfCorrection.cpp]: `validateInputs` is called by the automatically generated algorithm dialogs to check if the inputs are OK before permitting the actual execution of the algorithm.

File: Framework/Algorithms/src/HyspecScharpfCorrection.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (soininen) [reply Framework/Algorithms/src/HyspecScharpfCorrection.cpp]: I misinterpreted the comment. Sorry for the noise!

File: docs/source/algorithms/HyspecScharpfCorrection-v1.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (soininen) [reply docs/source/algorithms/HyspecScharpfCorrection-v1.rst]: If you think the users will understand the difference, then I think the current formula is OK.

File: Framework/Algorithms/src/HyspecScharpfCorrection.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (soininen) [reply Framework/Algorithms/src/HyspecScharpfCorrection.cpp]: Then I guess it's better to leave it like that and optimize it only if it proves to be a performance bottleneck.

File: Framework/Algorithms/src/HyspecScharpfCorrection.cpp [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+  } else {
+    throw std::invalid_argument(
+        "No Ei value has been set or stored within the run information.");
+  }
+
+  // Check if it is an event workspace
+  Mantid::DataObjects::EventWorkspace_const_sptr eventW =
+      boost::dynamic_pointer_cast<const Mantid::DataObjects::EventWorkspace>(
+          inputWS);
+  if (eventW != nullptr) {
```
[the remark this answers is not in the document]
Comment (AndreiSavici) [reply Framework/Algorithms/src/HyspecScharpfCorrection.cpp]: I still keep nullptr for type safety

File: docs/source/algorithms/HyspecScharpfCorrection-v1.rst [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+.. alias::
+
+.. properties::
+
+Description
+-----------
+
+This algorithm is applying a polarization correction for single crystal
+inelastic experiments. If one measures scattering intensity with polarization 
+along the momentum transfer  :math:`Q`, perpendicular to it in the horizontal
```
[the remark this answers is not in the document]
Comment (AndreiSavici) [reply docs/source/algorithms/HyspecScharpfCorrection-v1.rst]: This looks fine to me

[the remark this answers is not in the document]
Comment (soininen) [reply docs/source/algorithms/HyspecScharpfCorrection-v1.rst]: I meant to say there is two spaces before :math:. I think it is a typo.

**Review by soininen — APPROVED** [review]

**Comment by AndreiSavici** [comment]
Minor documentation rendering issues will be fixed separately.

**Pull request merged**